compiler/src/dotty/tools/dotc/ast/Trees.scala

```diff
@@ -1379,11 +1379,11 @@ object Trees {
       // Ties the knot of the traversal: call `foldOver(x, tree))` to dive in the `tree` node.
       def apply(x: X, tree: Tree)(using Context): X
 
-      def apply(x: X, trees: List[Tree])(using Context): X = trees match
-        case tree :: rest =>
-          apply(apply(x, tree), rest)
-        case Nil =>
-          x
+      def apply(x: X, trees: List[Tree])(using Context): X =
+        def fold(x: X, trees: List[Tree]): X = trees match
+          case tree :: rest => fold(apply(x, tree), rest)
+          case Nil => x
+        fold(x, trees)
```

Comment:
In ee4b125, I'm trying:

```Scala
        var acc = x
        var list = trees
        while (!list.isEmpty) do
          acc = apply(acc, list.head)
          list = list.tail
        acc
```

After tail-call optimization, the two versions are almost the same. Let's see if there is difference in benchmarks.

Reply:
I don't think there will be. If anything, the tail recursive version should be faster since it does a single type test per iteration instead of one each in isEmpty, head, and tail.

Reply:
In #9565, we get a slight speedup for Dotty.